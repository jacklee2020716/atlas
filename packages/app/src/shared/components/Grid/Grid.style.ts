import { StyleFn, makeStyles } from '../../utils'

export type GridStyleProps = {
  minItemWidth?: string | number
  maxItemWidth?: string | number
}
const container: StyleFn = (_, { minItemWidth = '300', maxItemWidth }) => ({
  display: 'grid',
  gridTemplateColumns: `repeat(auto-fit, minmax(${minItemWidth}px, ${maxItemWidth ? `${maxItemWidth}px` : '1fr'}))`,
  gap: '30px',
})

const item: StyleFn = () => ({
  cursor: 'pointer',
  width: '100%',
})

export const useCSS = (props: GridStyleProps) => ({
  container: makeStyles([container])(props),
  item: makeStyles([item])(props),
})
