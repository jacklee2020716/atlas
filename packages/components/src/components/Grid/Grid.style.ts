import { css } from "@emotion/core"

export type GridStyleProps = {
  minItemWidth?: string | number
  maxItemWidth?: string | number
}

export let makeStyles = ({
  minItemWidth = "300",
  maxItemWidth
}: GridStyleProps) => {
  return {
    container: css`
      display: grid;
      grid-template-columns: repeat(auto-fit, minmax(${minItemWidth}px, ${maxItemWidth ? maxItemWidth + "px" : "1fr"}));
      gap: 30px;
    `,
    item: css`
      width: 100%;
      cursor: pointer;
    `,
  }
}
