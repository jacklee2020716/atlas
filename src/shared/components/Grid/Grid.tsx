import React from 'react'
import styled from '@emotion/styled'
import useResizeObserver from 'use-resize-observer'
import { useMediaQuery } from 'react-responsive'
import { spacing } from '../../theme'

const toPx = (n: number | string) => (typeof n === 'number' ? `${n}px` : n)

const LARGE_VIEWPORT_BREAKPOINT = toPx(2000)

type ContainerProps = {
  gap: number | string
  gridTemplateColumns: string
}

const Container = styled.div<ContainerProps>`
  display: grid;
  gap: ${(props) => toPx(props.gap)};
  grid-template-columns: ${(props) => props.gridTemplateColumns};
`

type GridProps = {
  gap?: number | string
  className?: string
  maxColumns?: number
  minWidth?: number | string
  repeat?: 'fit' | 'fill'
  onResize?: (sizes: number[]) => void
}

const Grid: React.FC<GridProps> = ({
  className,
  gap = spacing.xl,
  onResize,
  repeat = 'fit',
  maxColumns = 6,
  minWidth = 300,
  ...props
}) => {
  const { ref: gridRef } = useResizeObserver<HTMLDivElement>({
    onResize: () => {
      if (onResize && gridRef.current) {
        const computedStyles = window.getComputedStyle(gridRef.current)
        const columnSizes = computedStyles.gridTemplateColumns.split(' ').map(parseFloat)
        onResize(columnSizes)
      }
    },
  })
  const isLargeViewport = useMediaQuery({ query: `(min-width: ${LARGE_VIEWPORT_BREAKPOINT})` })

  const gridTemplateColumns = isLargeViewport
    ? `repeat(${maxColumns}, 1fr)`
    : `repeat(auto-${repeat}, minmax(min(${toPx(minWidth)}, 100%), 1fr))`

  return (
    <Container {...props} className={className} ref={gridRef} gap={gap} gridTemplateColumns={gridTemplateColumns} />
  )
}
export default Grid
