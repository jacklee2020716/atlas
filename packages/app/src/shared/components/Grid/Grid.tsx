import React from 'react'
import { GridStyleProps, useCSS } from './Grid.style'

type SectionProps = {
  items?: React.ReactNode[]
  className?: string
} & GridStyleProps

export default function Grid({ items = [], className = '', ...styleProps }: SectionProps) {
  let styles = useCSS(styleProps)
  return (
    <div css={styles.container} className={className}>
      {items.map((item, index) => (
        <div key={`grid-item-${index}`} css={styles.item}>
          {item}
        </div>
      ))}
    </div>
  )
}
