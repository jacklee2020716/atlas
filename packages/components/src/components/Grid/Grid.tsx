import React from "react"
import { makeStyles, GridStyleProps } from "./Grid.style"

type SectionProps = {
  items?: React.ReactNode[]
  className?: string
} & GridStyleProps

export default function Grid({
  items = [],
  className = "",
  ...styleProps
}: SectionProps) {
  let styles = makeStyles(styleProps)
  return (
    <div css={styles.container} className={className}>
      {items.map((item, index) => <div key={`grid-item-${index}`} css={styles.item}>{item}</div>)}
    </div>
  )
}
