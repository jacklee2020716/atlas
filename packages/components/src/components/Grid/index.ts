import { memo } from "react"
import Grid from "./Grid"
export default memo(Grid)
