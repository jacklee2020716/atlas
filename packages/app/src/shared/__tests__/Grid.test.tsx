import React from 'react'
import { shallow } from 'enzyme'
import Grid from '@/shared/components/Grid'
describe('Grid component', () => {
  it('Should render Grid correctly', () => {
    expect(
      shallow(
        <Grid>
          <div>Grid Elemen 1</div>
          <div>Grid Elemen 2</div>
          <div>Grid Elemen 3</div>
        </Grid>
      )
    ).toMatchSnapshot()
  })
})
