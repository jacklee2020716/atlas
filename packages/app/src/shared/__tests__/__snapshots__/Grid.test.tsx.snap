// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Grid component Should render Grid correctly 1`] = `ShallowWrapper {}`;
